src/analyses/cfg_dominators.h

```diff
@@ -16,20 +16,163 @@ Author: Georg Weissenbacher, georg@weissenbacher.name
 #include <list>
 #include <map>
 #include <iosfwd>
-#include <cassert>
+#include <algorithm>
+#include <fstream>
+#include <stack>
+#include <tuple>
+#include <type_traits>
+#include <iterator>
 
 #include <goto-programs/goto_functions.h>
 #include <goto-programs/goto_program.h>
 #include <goto-programs/cfg.h>
 
-template <class P, class T, bool post_dom>
-class cfg_dominators_templatet
-{
-public:
-  typedef std::set<T> target_sett;
+template<typename T, typename node_indext>
+struct dominators_datat {
+  std::vector<T> data;
+  std::vector<node_indext> immediate_dominator;
+};
+
+/// An immutable set of dominators. Constant memory usage and creation time, but linear lookup time
+/// Immutability is necessary because the structure uses sharing
+template<typename T, typename node_indext>
+struct dominatorst {
+  using datat = dominators_datat<T, node_indext>;
+  static constexpr node_indext npos = std::numeric_limits<node_indext>::max();
+
+  /// Create an empty set
+  /// Note: Only unreachable nodes should be assigned empty sets after the algorithm completes
+  dominatorst() : dominators_data(nullptr), node_index(npos), cached_distance(0) {
+  }
+
+  /// Create the dominators set of node_index
+  dominatorst(datat &dominators_data, node_indext node_index)
+      : dominators_data(&dominators_data),
+        node_index(node_index),
+        cached_distance(npos) {
+  }
+
+  dominatorst(const dominatorst &other)
+      : dominators_data(other.dominators_data),
+        node_index(other.node_index),
+        cached_distance(other.cached_distance) {
+  }
+
+  dominatorst &operator=(const dominatorst &rhs) {
+    dominators_data = rhs.dominators_data;
+    node_index = rhs.node_index;
+    cached_distance = rhs.cached_distance;
+    return *this;
+  }
+
+  datat *dominators_data;
+  node_indext node_index;
+  mutable std::size_t cached_distance;
```

Comment:
Shouldn't this be a class with those three being made `private`? Certainly true for a pointer.

Reply:
There's no reason for it not to be, so sure I can change this.